atom/browser/web_dialog_helper.cc

```diff
@@ -71,23 +72,61 @@ class FileSelectHelper : public base::RefCounted<FileSelectHelper>,
             isolate, base::Bind(&FileSelectHelper::OnSaveDialogDone, this)))));
   }
 
+  void OnDirectoryListerDone(std::vector<FileChooserFileInfoPtr> file_info,
+                             base::FilePath base_dir) override {
+    OnFilesSelected(std::move(file_info), base_dir);
+    Release();
+  }
+
  private:
   friend class base::RefCounted<FileSelectHelper>;
 
   ~FileSelectHelper() override {}
 
+  void EnumerateDirectory(base::FilePath base_dir) {
+    auto* lister = new net::DirectoryLister(
+        base_dir, net::DirectoryLister::NO_SORT_RECURSIVE,
+        new atom::DirectoryListerHelper(base_dir, this));
+    lister->Start();
+    // It is difficult for callers to know how long to keep a reference to
+    // this instance.  We AddRef() here to keep the instance alive after we
+    // return to the caller.  Once the directory lister is complete we
+    // Release() in OnDirectoryListerDone() and at that point we run
+    // OnFilesSelected() which will deref the last reference held by the
+    // listener.
+    AddRef();
+  }
+
   void OnOpenDialogDone(mate::Dictionary result) {
     std::vector<FileChooserFileInfoPtr> file_info;
     bool canceled = true;
     result.Get("canceled", &canceled);
+    base::FilePath base_dir;
+    // For certain file chooser modes (kUploadFolder) we need to do some async
+    // work before calling back to the listener.  In that particular case the
+    // listener is called from the directory enumerator.
+    bool ready_to_call_listener = false;
 
     if (!canceled) {
       std::vector<base::FilePath> paths;
       if (result.Get("filePaths", &paths)) {
-        for (auto& path : paths) {
-          file_info.push_back(FileChooserFileInfo::NewNativeFile(
-              blink::mojom::NativeFileInfo::New(
-                  path, path.BaseName().AsUTF16Unsafe())));
+        // If we are uploading a folder we need to enumerate its contents
+        if (mode_ == FileChooserParams::Mode::kUploadFolder &&
+            paths.size() >= 1) {
+          base_dir = paths[0];
+
+          // Actually enumerate soemwhere off-thread
+          base::SequencedTaskRunnerHandle::Get()->PostTask(
```

Comment:
This doesn't mean off-thread, it will be a task that runs on the same sequence as where its called but will be guaranteed to run in a defined order. Prefer `base::PostTaskWithTraits`.